uat/custom-components/src/main/java/com/aws/greengrass/artifacts/ShadowReactiveComponent.java

```diff
@@ -0,0 +1,233 @@
+            String updateShadowDocumentRequest1 = args[2];
+            byte[] updateShadowDocumentPayload1 = null;
+            JsonNode expectedUpdateShadowDocumentJson = null;
+            JsonNode expectedUpdateShadowDocumentJson2 = null;
+            if (!isNullOrEmpty(updateShadowDocumentRequest1)) {
+                updateShadowDocumentPayload1 = updateShadowDocumentRequest1.getBytes(UTF_8);
+            }
+            String expectedUpdateShadowDocumentRequest1 = args[3];
+            if (!isNullOrEmpty(expectedUpdateShadowDocumentRequest1)) {
+                byte[] expectedUpdateShadowDocumentPayload1 = expectedUpdateShadowDocumentRequest1.getBytes(UTF_8);
+                expectedUpdateShadowDocumentJson = MAPPER.readTree(expectedUpdateShadowDocumentPayload1);
+            }
+            if (updateShadowDocumentPayload1 != null) {
+                handleUpdateThingShadowOperation(thingName, shadowName, updateShadowDocumentPayload1, expectedUpdateShadowDocumentJson);
+            }
+
+            String subscribeTopic = System.getenv("SubscribeTopic");
+            if (subscribeTopic != null) {
+                String expectedUpdateShadowDocumentRequest2 = args[4];
+                if (!isNullOrEmpty(expectedUpdateShadowDocumentRequest2)) {
+                    byte[] expectedUpdateShadowDocumentPayload2 = expectedUpdateShadowDocumentRequest2.getBytes(UTF_8);
+                    if (expectedUpdateShadowDocumentPayload2.length > 0) {
+                        expectedUpdateShadowDocumentJson2 = MAPPER.readTree(expectedUpdateShadowDocumentPayload2);
+                    }
+                }
+                subscribeToShadowUpdates(thingName, shadowName, subscribeTopic, expectedUpdateShadowDocumentJson, expectedUpdateShadowDocumentJson2);
+            }
+        } catch (InterruptedException | ExecutionException | IOException e) {
+            LOGGER.error("Error", e);
+            System.exit(1);
+        } finally {
+            if (eventStreamRpcConnection != null) {
+                try {
+                    eventStreamRpcConnection.close();
+                } catch (Exception ex) {
+                    LOGGER.error("Unexpected error occurred while closing IPC connection", ex);
+                }
+            }
+        }
+    }
+
+    private boolean isNullOrEmpty(String s) {
+        return s == null || "".equals(s) || "null".equals(s);
+    }
+
+    private void updateComponentState(GreengrassCoreIPC greengrassCoreIPCClient, ReportedLifecycleState state) throws InterruptedException, ExecutionException {
+        UpdateStateRequest updateStateRequest = new UpdateStateRequest();
+        updateStateRequest.setState(state);
+        greengrassCoreIPCClient.updateState(updateStateRequest, Optional.empty()).getResponse().get();
+    }
+
+    private void subscribeToShadowUpdates(String thingName, String shadowName, String topic, JsonNode expectedUpdateShadowDocumentJson, JsonNode expectedUpdateShadowDocumentJson2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.